tutorials/tour/_posts/2017-02-13-generic-classes.md

```diff
@@ -9,40 +9,50 @@ categories: tour
 num: 18
 next-page: variances
 previous-page: sequence-comprehensions
+assumed-knowledge: classes
 ---
+Generic classes are classes which take a type as a parameter. They are particularly useful for collection classes.
 
-Like in Java 5 (aka. [JDK 1.5](http://java.sun.com/j2se/1.5/)), Scala has built-in support for classes parameterized with types. Such generic classes are particularly useful for the development of collection classes.
-Here is an example which demonstrates this:
-
+## Defining a generic classes
+Generic classes take a type as a parameter within square brackets `[]`. One convention is to use the letter `T` as type parameter identifier, though this is not required.
 ```tut
 class Stack[T] {
-  var elems: List[T] = Nil
-  def push(x: T) { elems = x :: elems }
-  def top: T = elems.head
-  def pop() { elems = elems.tail }
+  private var elements: List[T] = Nil
+  def push(x: T) { elements = x :: elements }
+  def peek: T = elements.head
+  def pop(): T = {
+    val currentTop = peek
+    elements = elements.tail
+    currentTop
+  }
 }
 ```
+This implementation of a `Stack` class takes any type `T` as a parameter. This means the underlying list, `var elems: List[T] = Nil`, can only store elements of type `T`. The procedure `def push` only accepts objects of type `T` (note: `elems = x :: elems` reassigns `elems` to a new list created by prepending `x` to the current `elems`).
 
-Class `Stack` models imperative (mutable) stacks of an arbitrary element type `T`. The type parameters enforces that only legal elements (that are of type `T`) are pushed onto the stack. Similarly, with type parameters we can express that method `top` will only yield elements of the given type.
-
-Here are some usage examples:
+## Usage
 
-```tut
-object GenericsTest extends App {
-  val stack = new Stack[Int]
-  stack.push(1)
-  stack.push('a')
-  println(stack.top)
-  stack.pop()
-  println(stack.top)
-}
+To use a generic class, put the type in the square brackets in place of `T`.
 ```
+val stack = new Stack[Int]
+stack.push(1)
+stack.push('a')
+println(stack.peek)  // prints 97
+println(stack.pop)  // prints 97
+println(stack.peek)  // prints 1
+```
+Here we have a stack of Ints. Notice how we can pass in the Char object `a`. This is because the `Char` class has a [method to implicitly convert a `Char` to an Int](http://scala-lang.org/api/current/scala/Char$.html#char2int(x:Char):Int).
```

Comment:
Is this worth to be explained here? I think the example adds too much info to the reader, which is assumed to be familiar with implicit conversions. I would just remove the char example :smile:.